gh-100107: Make Py launcher ignore app aliases that launch Microsoft Store

flagrama opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->
Currently when you use the `#!/usr/bin/env python3` shebang on Windows 10 or Windows 11 the Py launcher will search the PATH environment variable and find the default NTFS reparse points called App Execution Aliases Microsoft added to the OSes. These are effectively stub exes that, when Python is not installed from the Microsoft Store, will launch the Microsoft Store to install Python, or if launched with arguments will report an error. 

This makes Windows 10/11 the only OS where the standard shebang does not work. It complicated packaging python scripts for ordinary users because they have an extra step, either disabling the app execution aliases, installing from the Microsoft Store, or launching the script manually. Windows ends up being the only OS that doesn't "just work" and changing the shebang to something else will break or degrade the experience in other OSes.

This PR solves this issue by checking if the exe found in the PATH is a reparse point. If it is, it will read the file and identify the alias that will launch the store and will drop out of the shebang handling as if nothing was found in the PATH. If it is an alias that does not launch the store, or an ordinary file it will just continue and use the found file as there shouldn't be any issues.

In my testing this now makes opening a script with the standard `#!/usr/bin/env python3` shebang work without additional work by the user.

Note that I am not a Windows user or a Windows developer so it's possible that I've missed some workflow and don't know if it works, or my code may be inefficient. I've only really tested the workflow of the normal users that use a Python application I help develop and know that I have solved the issue there.

<!-- gh-issue-number: gh-100107 -->
* Issue: gh-100107
<!-- /gh-issue-number -->

flagrama committed:
pylauncher: Ignore shebang if APPEXECLINK found in PATH

flagrama committed:
pylauncher: Only ignore alias that opens app installer

flagrama committed:
pylauncher: add error handling

zooba commented on PC/launcher2.c:
```diff
@@ -826,6 +841,72 @@ searchPath(SearchInfo *search, const wchar_t *shebang, int shebangLength)
+        winerror(0, L"Failed to find %s on PATH\n", filename);
+        return RC_BAD_VIRTUAL_PATH;
+    }
+
```
Should be able to `FindClose` here and not worry about cleaning it up later.

zooba commented on PC/launcher2.c:
```diff
@@ -826,6 +841,72 @@ searchPath(SearchInfo *search, const wchar_t *shebang, int shebangLength)
         return RC_BAD_VIRTUAL_PATH;
     }
 
+    // Make sure we didn't find a reparse point that will open the Microsoft Store
```
Let's start by refactoring all this out into another function `ensure_no_redirector_stub` that either returns `0` (okay) or the RC constant that should be returned from here.

zooba commented on PC/launcher2.c:
```diff
@@ -826,6 +841,72 @@ searchPath(SearchInfo *search, const wchar_t *shebang, int shebangLength)
+    }
+
+    if (findData.dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT &&
+        findData.dwReserved0 & IO_REPARSE_TAG_APPEXECLINK) {
```
After refactoring, use this check to exit early from the new function. That will keep the overall indent level lower for the rest of it.

zooba commented on PC/launcher2.c:
```diff
@@ -826,6 +841,72 @@ searchPath(SearchInfo *search, const wchar_t *shebang, int shebangLength)
+                debug(L"# %s on disappeared\n", buffer);
+                // The file found on PATH disappeared. Alias probably disabled by user while trying to run, let normal handling take over
+                RC_NO_SHEBANG;
+            }
```
I don't think this error adds anything - we could lose this race at any time, so probably best to just return `RC_NO_SHEBANG` for them all and it can fail if/when we try to launch it.

zooba commented on PC/launcher2.c:
```diff
@@ -826,6 +841,72 @@ searchPath(SearchInfo *search, const wchar_t *shebang, int shebangLength)
+                debug(L"# %s on disappeared\n", buffer);
+                // The file found on PATH disappeared. Alias probably disabled by user while trying to run, let normal handling take over
+                RC_NO_SHEBANG;
+            }
```
Ditto - no benefit to calling this one out separately. Normal handling for all errors should be okay

zooba commented on PC/launcher2.c:
```diff
@@ -826,6 +841,72 @@ searchPath(SearchInfo *search, const wchar_t *shebang, int shebangLength)
+
+        appExecLinkFilePtr = (AppExecLinkFile*)&appExecLinkBuf;
+
+        wchar_t redirectorPackageId[] = L"Microsoft.DesktopAppInstaller";
```
```suggestion
        const wchar_t *redirectorPackageId = L"Microsoft.DesktopAppInstaller_8wekyb3d8bbwe";
```

zooba commented on PC/launcher2.c:
```diff
@@ -826,6 +841,72 @@ searchPath(SearchInfo *search, const wchar_t *shebang, int shebangLength)
+    if (findData.dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT &&
+        findData.dwReserved0 & IO_REPARSE_TAG_APPEXECLINK) {
+        const size_t bufSize = sizeof(AppExecLinkFile);
+        wchar_t appExecLinkBuf[sizeof(AppExecLinkFile)];
```
These aren't necessary - just allocate the struct:

```
AppExecLinkFile appExecLink;
```

And then use `sizeof(appExecLink)` anywhere we need it, which ought to only be the DeviceIOControl call.

zooba commented on PC/launcher2.c:
```diff
@@ -826,6 +841,72 @@ searchPath(SearchInfo *search, const wchar_t *shebang, int shebangLength)
+        wchar_t partialId[30];
+        wcsncpy_s(partialId, redirectorStrLen + 1, appExecLinkFilePtr->stringList, redirectorStrLen);
+
+        if (0 == wcsncmp(partialId, &redirectorPackageId, redirectorStrLen)) {
```
A simple `wcscmp` ought to be fine here if you use the full package name. As far as I'm aware, the publisher ID has never changed and never will (it's always some kind of hash of "Microsoft").

zooba reviewed (changes requested):
It's got all the pieces we need, though it can be simplified quite a bit. Would also be ideal to move it to its own function, so that someone reading this function can see from the call what it's doing but they don't have to actually read the details unless they want to.

flagrama committed:
pylauncher: Address code review

move new code to own function
simplify error handling
simplify struct handling
simplify string comparison

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

flagrama committed:
pylauncher: Move struct allocation closer to usage

flagrama committed:
Merge branch 'main' into py-launcher-shebang-appexeclink

flagrama committed:
pylauncher: adjust newlines

flagrama committed:
Merge branch 'py-launcher-shebang-appexeclink' of https://github.com/flagrama/cpython into py-launcher-shebang-appexeclink

zooba commented on PC/launcher2.c:
```diff
@@ -778,6 +778,54 @@ _shebangStartsWith(const wchar_t *buffer, int bufferLength, const wchar_t *prefi
+    if (!DeviceIoControl(hReparsePoint, FSCTL_GET_REPARSE_POINT, NULL, 0, &appExecLink, sizeof(appExecLink), NULL, NULL)) {
+        // Let normal handling take over
+        debug(L"# Did not find %s on PATH\n", filename);
+        return RC_NO_SHEBANG;
```
```suggestion
        CloseHandle(hReparsePoint);
        return RC_NO_SHEBANG;
```

flagrama committed:
pylauncher: cleanup leaked handle

Co-authored-by: Steve Dower <steve.dower@microsoft.com>
